### PR TITLE
AMM ReduceReplicas to iterate only on replicated tasks

### PR DESCRIPTION
**Pull request by crusaderky**
Encapsulate changes to ts._who_has / ws._has_what.
Track in-memory tasks with more than one replica apart from all the others.
Change the AMM ReduceReplicas policy to iterate only on this subset, which in the vast majority of the cases should be much smaller than the set of all tasks in the cluster.

CC @mrocklin @fjetter @jrbourbeau @gjoseph92

### CODE REVIEW

**Comment by crusaderky** [comment]
ping @fjetter @gjoseph92  - this has potential performance implications on the whole scheduler, so I'd very much like your review.

File: distributed/scheduler.py [hunk 1]
```diff
@@ -4733,70 +4770,33 @@ def stimulus_task_erred(
...
-
-            parent._transitions(recommendations, client_msgs, worker_msgs)
-            recommendations = {}
-
-            if parent._validate:
-                assert cause not in self.who_has
-
-            return recommendations, client_msgs, worker_msgs
+                return {}, {}, {}
+            return {key: "released"}, {}, {}
```
Comment (fjetter) [inline distributed/scheduler.py:4799]: I found myself more than once in a situation where I wanted to delete this stimulus handler since I believe most of it is dead code. However, I don't see the connection to the current PR. Is this somehow related? Can this change be split off?

File: distributed/scheduler.py [hunk 2]
```diff
@@ -6751,17 +6741,14 @@ def update_data(
...
-                else:
-                    ts_nbytes = ts.get_nbytes()
+
                 for w in workers:
                     ws: WorkerState = parent._workers_dv[w]
-                    if ts not in ws._has_what:
-                        ws._nbytes += ts_nbytes
-                        ws._has_what[ts] = None
-                        ts._who_has.add(ws)
+                    if ws not in ts._who_has:
```
Comment (fjetter) [inline distributed/scheduler.py:6750]: I noticed changes like these in a few different places. You replaced a `ts not in ws._has_what` with a `ws not in ts._who_has`. Assuming the state machine is not corrupt, these are equivalent statements. What motivated this change?

**Review by fjetter — COMMENTED** [review]
LGTM. I don't have any performance concerns

File: distributed/scheduler.py (hunk 2 above)
Comment (crusaderky) [reply distributed/scheduler.py:6750]: pure aesthetics and consistency. Nothing beyond that.

File: distributed/scheduler.py (hunk 1 above)
Comment (crusaderky) [reply distributed/scheduler.py:4799]: Actually I can't. If you read the old code, it does something clearly wrong - on lines 4783:4787, it deletes ts from has_what but it removes ``cts.nbytes`` from ``ws.nbytes``; it should have removed ``ts.nbytes`` instead. With nothing invoking this code path, I had no idea what the original intent of the logic was, so I opted for cutting the method down to just what is actually invoked by ``handle_release_data``.

Comment (crusaderky) [reply distributed/scheduler.py:4799]: if you prefer I can remove the ``stimulus_missing_data`` method altogether and inline the surviving code into ``handle_release_data``?

File: distributed/scheduler.py [hunk 3]
```diff
@@ -3425,6 +3429,39 @@ def worker_objective(self, ts: TaskState, ws: WorkerState) -> tuple:
...
+
+        ws._nbytes += ts.get_nbytes()
+        ws._has_what[ts] = None
+        ts._who_has.add(ws)
+        if len(ts._who_has) == 2:
+            self._replicated_tasks.add(ts)
+
+    @ccall
+    def remove_replica(self, ts: TaskState, ws: WorkerState):
+        """Note that a worker no longer holds a replica of a task"""
```
Comment (gjoseph92) [inline distributed/scheduler.py:3447]: Nit: maybe add a `if self._validate` case to this and `remove_all_replicas` to catch any issues in future tests?

File: distributed/scheduler.py (hunk 1 above)
Comment (gjoseph92) [reply distributed/scheduler.py:4799]: Maybe that's reasonable, since it's not really a `stimulus` function anymore either and it's only a few lines.

File: distributed/scheduler.py [hunk 4]
```diff
@@ -3425,6 +3429,39 @@ def worker_objective(self, ts: TaskState, ws: WorkerState) -> tuple:
...
             return (start_time, ws._nbytes)
 
+    @ccall
+    def add_replica(self, ts: TaskState, ws: WorkerState):
+        """Note that a worker holds a replica of a task with state='memory'"""
+        if self._validate:
+            assert ws not in ts._who_has
+            assert ts not in ws._has_what
+
+        ws._nbytes += ts.get_nbytes()
```
Comment (gjoseph92) [inline distributed/scheduler.py:3439]: Perhaps some mysterious magic type annotations are necessary to get this function to Cythonize well? I have no idea.

**Review by gjoseph92 — COMMENTED** [review]
LGTM as well. Nice to have this consolidated into a function anyway.

File: distributed/scheduler.py (hunk 1 above)
Comment (fjetter) [reply distributed/scheduler.py:4799]: I'm fine with moving the code to `handle_release_data`

File: distributed/scheduler.py (hunk 3 above)
Comment (crusaderky) [reply distributed/scheduler.py:3447]: both ``remove_replica`` and ``remove_all_replicas`` already raise KeyError

File: distributed/scheduler.py (hunk 1 above)
Comment (crusaderky) [reply distributed/scheduler.py:4799]: done; please double check that the new code is functionally identical to the previous one

File: distributed/scheduler.py (hunk 4 above)
Comment (crusaderky) [reply distributed/scheduler.py:3439]: Could you elaborate on why you think it's not cythonizing well? I'm trying to build the HTML report but failing:

```

$ python -m pip install -vv --no-deps --install-option="--with-cython" -e .
(works, but doesn't generate the HTML)
$ cythonize -a distributed/scheduler.py
    @ccall
    @inline
    @nogil
    def get_nbytes(self) -> Py_ssize_t:
                  ^
------------------------------------------------------------
distributed/scheduler.py:1638:19: Operation not allowed without gil
```

Comment (crusaderky) [reply distributed/scheduler.py:3439]: Added explicit declaration of nbytes.
I also noticed that replacing 
```python
        ws._nbytes -= ts.get_nbytes()
        del ws._has_what[ts]
        ts._who_has.remove(ws)
        if len(ts._who_has) == 1:
            self._replicated_tasks.remove(ts)
```
with
```python
        wh: set = ts._who_has
        hw: dict = ws._has_what
        ws._nbytes -= ts.get_nbytes()
        del hw[ts]
        wh.remove(ws)
        if len(wh) == 1:
            rt: set = self._replicated_tasks
            rt.remove(ts)
```
marginally improves the Cython code, at the cost of greatly reduced readability, so I'd rather not do it.

**Comment by crusaderky** [comment]
All code review critiques have been incorporated. The new ``stimulus_task_erred`` and ``handle_release_data`` are supposed to be functionally identical to before but please double check.

**Review by fjetter — APPROVED** [review]
I reviewed the changes around handle_missing_data / stimulus_missing_data again and agree that they should be identical. The missing-data system evolved quite dramatically over time and I also discovered several places on worker side as well where the code does no longer make a lot of sense. I believe this is genuinely dead code.

Test failures appear to be unrelated as well
